org.h2.mvstore.MVStoreException: Chunk metadata too long

This happend when I tried to operate on my 214 DB with the latest master:

    org.h2.jdbc.JdbcSQLNonTransientException: (Message HY000 not found) [50000-219]

    	at org.h2.message.DbException.getJdbcSQLException(DbException.java:554)

    	at org.h2.message.DbException.getJdbcSQLException(DbException.java:477)

    	at org.h2.message.DbException.get(DbException.java:212)

    	at org.h2.engine.Database.throwLastBackgroundException(Database.java:1783)

    	at org.h2.engine.SessionLocal.beforeCommitOrRollback(SessionLocal.java:761)

    	at org.h2.engine.SessionLocal.commit(SessionLocal.java:675)

    	at org.h2.command.Command.commitIfNonTransactional(Command.java:309)

    	at org.h2.command.Command.executeUpdate(Command.java:243)

    	at
        org.h2.jdbc.JdbcPreparedStatement.executeUpdateInternal(JdbcPreparedStatement.java:209)

    	at org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:169)

    	at matdb.storage.Cache.remove(Cache.java:424)

    Caused by: org.h2.message.DbException: (Message HY000 not found) [50000-219]

    	at org.h2.message.DbException.get(DbException.java:212)

    	at org.h2.message.DbException.convert(DbException.java:395)

    	at org.h2.mvstore.db.Store.lambda$new$0(Store.java:122)

    	at org.h2.mvstore.MVStore.handleException(MVStore.java:1556)

    	at org.h2.mvstore.MVStore.panic(MVStore.java:371)

    	at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1518)

    	at org.h2.mvstore.FileStore.lambda$serializeAndStore$3(FileStore.java:1422)

    	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)

    	at java.util.concurrent.FutureTask.run(FutureTask.java:266)

    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

    	at java.lang.Thread.run(Thread.java:748)

    Caused by: org.h2.jdbc.JdbcSQLNonTransientException: (Message HY000 not found)
    [50000-219]

    	at org.h2.message.DbException.getJdbcSQLException(DbException.java:554)

    	at org.h2.message.DbException.getJdbcSQLException(DbException.java:477)

    	... 12 more

    Caused by: org.h2.mvstore.MVStoreException: Chunk metadata too long 131 133
    chunk:22694,len:d99,pages:604,pinCount:199,max:fc98a0,map:199,root:89a50035b21e0b,time:800e9db3,version:127e10,next:923df3,toc:d95321
    [2.2.219/3]

    	at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)

    	at org.h2.mvstore.Chunk.writeChunkHeader(Chunk.java:252)

    	at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1510)

    	... 6 more

--- a/h2/src/main/org/h2/mvstore/Chunk.java
+++ b/h2/src/main/org/h2/mvstore/Chunk.java
@@ -32,11 +32,10 @@ public abstract class Chunk<C extends Chunk<C>> {
 
     /**
      * The maximum length of a chunk header, in bytes.
-     * chunk:ffffffff,len:ffffffff,pages:ffffffff,pinCount:ffffffff,map:ffffffff,
-     * root:ffffffffffffffff,time:ffffffffffffffff,toc:ffffffff,version:ffffffffffffffff,
-     * next:ffffffffffffffff
+     * chunk:ffffffff,len:ffffffff,pages:ffffffff,pinCount:ffffffff,max:ffffffffffffffff,map:ffffffff,
+     * root:ffffffffffffffff,time:ffffffffffffffff,version:ffffffffffffffff,next:ffffffffffffffff,toc:ffffffff
      */
-    static final int MAX_HEADER_LENGTH = 1024;
+    static final int MAX_HEADER_LENGTH = 1024;  // 199 really
 
     /**
      * The length of the chunk footer. The longest footer is:
@@ -239,20 +238,40 @@ public abstract class Chunk<C extends Chunk<C>> {
     /**
      * Write the chunk header.
      *
-     * @param buff the target buffer
-     * @param minLength the minimum length
+     * @return estimated size of the header
      */
-    void writeChunkHeader(WriteBuffer buff, int minLength) {
-        long delimiterPosition = buff.position() + minLength - 1;
-        buff.put(getHeaderBytes());
+    int estimateHeaderSize() {
+        byte[] headerBytes = getHeaderBytes();
+        int headerLength = headerBytes.length;
+        // Initial chunk will look like (length-wise) something in between those two lines:
+        // chunk:0,len:0,pages:0,max:0,map:0,root:0,time:0,version:0                                      // 57
+        // chunk:ffffffff,len:0,pages:0,max:0,map:0,root:0,time:ffffffffffffffff,version:ffffffffffffffff // 94
+        assert 57 <= headerLength && headerLength <= 94 : headerLength + " " + getHeader();
+        // When header is fully formed, it will grow and here are fields,
+        // which do not exist in initial header or may grow from their initial values:
+        // len:0[fffffff],pages:0[fffffff][,pinCount:ffffffff],max:0[fffffffffffffff],map:0[fffffff],
+        // root:0[fffffffffffffff,next:ffffffffffffffff,toc:fffffffff]                       // 104 extra chars
+        return headerLength + 104 + 1; // extra one for the terminator
+    }
+
+    /**
+     * Write the chunk header.
+     *
+     * @param buff the target buffer
+     * @param maxLength length of the area reserved for the header
+     */
+    void writeChunkHeader(WriteBuffer buff, int maxLength) {
+        long delimiterPosition = buff.position() + maxLength - 1;
+        byte[] headerBytes = getHeaderBytes();
+        buff.put(headerBytes);
         while (buff.position() < delimiterPosition) {
             buff.put((byte) ' ');
         }
-        if (minLength != 0 && buff.position() > delimiterPosition) {
+        if (maxLength != 0 && buff.position() > delimiterPosition) {
             throw DataUtils.newMVStoreException(
                     DataUtils.ERROR_INTERNAL,
                     "Chunk metadata too long {0} {1} {2}", delimiterPosition, buff.position(),
-                    new String(getHeaderBytes(), StandardCharsets.ISO_8859_1));
+                    getHeader());
         }
         buff.put((byte) '\n');
     }
@@ -339,6 +358,10 @@ public abstract class Chunk<C extends Chunk<C>> {
             DataUtils.appendMap(buff, ATTR_OCCUPANCY,
                     StringUtils.convertBytesToHex(occupancy.toByteArray()));
         }
+    }
+
+    public String getHeader() {
+        return new String(getHeaderBytes(), StandardCharsets.ISO_8859_1);
     }
 
     private byte[] getHeaderBytes() {

--- a/h2/src/main/org/h2/mvstore/Chunk.java
+++ b/h2/src/main/org/h2/mvstore/Chunk.java
@@ -261,16 +261,16 @@ public abstract class Chunk<C extends Chunk<C>> {
      * @param maxLength length of the area reserved for the header
      */
     void writeChunkHeader(WriteBuffer buff, int maxLength) {
-        long delimiterPosition = buff.position() + maxLength - 1;
+        int terminatorPosition = buff.position() + maxLength - 1;
         byte[] headerBytes = getHeaderBytes();
         buff.put(headerBytes);
-        while (buff.position() < delimiterPosition) {
+        while (buff.position() < terminatorPosition) {
             buff.put((byte) ' ');
         }
-        if (maxLength != 0 && buff.position() > delimiterPosition) {
+        if (maxLength != 0 && buff.position() > terminatorPosition) {
             throw DataUtils.newMVStoreException(
                     DataUtils.ERROR_INTERNAL,
-                    "Chunk metadata too long {0} {1} {2}", delimiterPosition, buff.position(),
+                    "Chunk metadata too long {0} {1} {2}", terminatorPosition, buff.position(),
                     getHeader());
         }
         buff.put((byte) '\n');

--- a/h2/src/main/org/h2/mvstore/FileStore.java
+++ b/h2/src/main/org/h2/mvstore/FileStore.java
@@ -1436,8 +1436,7 @@ public abstract class FileStore<C extends Chunk<C>>
 
     private void serializeToBuffer(WriteBuffer buff, ArrayList<Page<?, ?>> changed, C c, C previousChunk) {
         // need to patch the header later
-        c.writeChunkHeader(buff, 0);
-        int headerLength = buff.position() + 58; // len:0[fffffff]map:0[fffffff],toc:0[ffffffff],root:0[fffffffffffffff,next:ffffffffffffffff]
+        int headerLength = c.estimateHeaderSize();
         buff.position(headerLength);
         c.next = headerLength;
 

